lib/s3routes/routesUtils.ts

```diff
@@ -441,23 +456,38 @@ function retrieveData(
                 });
                 // errors on server side with readable stream
                 readable.on('error', err => {
-                    log.error('error piping data from source');
+                    log.error('error piping data from source', { error: err });
                     _destroyResponse();
                     return cbOnce(err);
                 });
+                // At this point, we assume that the data backend is available.
+                // We write the http headers before piping the data to the
+                // response object.
+                // Any subsequent error leads to destror deletion.
+                if (!response.headersSent) {
+                    response.writeHead(httpHeadersOnSucess.statusCode, httpHeadersOnSucess.headers);
+                }
                 currentStream = readable;
                 return readable.pipe(response, { end: false });
             }), err => {
-            currentStream = null;
-            if (err) {
-                log.debug('abort response due to error', {
-                    // @ts-expect-error
-                    error: err.code, errMsg: err.message });
-            }
-            // call end for all cases (error/success) per node.js docs
-            // recommendation
-            response.end();
-        },
+                currentStream = null;
+                if (err) {
+                    log.error('abort response due to error', {
+                        // @ts-expect-error
+                        error: err.code, errMsg: err.message });
+                    // Only end if we haven't already destroyed the response
+                    if (!responseDestroyed) {
```

Comment:
should we not just check `!response.headersSent` instead?

Reply:
Yes, code is updated.